File: modules/debugger/src/main/java/com/falsepattern/zigbrains/clion/ZigClionDebuggerDriverConfigurationProvider.java
/*
 * Copyright 2023-2024 FalsePattern
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.clion;

import com.falsepattern.zigbrains.debugbridge.ZigDebuggerDriverConfigurationProvider;
import com.falsepattern.zigbrains.debugger.settings.ZigDebuggerSettings;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.SystemInfo;
import com.jetbrains.cidr.cpp.execution.debugger.backend.CLionGDBDriverConfiguration;
import com.jetbrains.cidr.cpp.execution.debugger.backend.CLionLLDBDriverConfiguration;
import com.jetbrains.cidr.cpp.toolchains.CPPToolchains;
import com.jetbrains.cidr.execution.debugger.backend.DebuggerDriverConfiguration;
import lombok.val;
import org.jetbrains.annotations.Nullable;

import java.util.function.Supplier;

public class ZigClionDebuggerDriverConfigurationProvider implements ZigDebuggerDriverConfigurationProvider {
    private static final Logger LOG = Logger.getInstance(ZigClionDebuggerDriverConfigurationProvider.class);

    @Override
    public @Nullable Supplier<DebuggerDriverConfiguration> getDebuggerConfiguration(Project project, boolean isElevated, boolean emulateTerminal) {
        if (SystemInfo.isWindows)
            return null;

        if (!ZigDebuggerSettings.getInstance().useClion)
            return null;

        val toolchains = CPPToolchains.getInstance();
        var toolchain = toolchains.getToolchainByNameOrDefault("Zig");
        if (toolchain == null || !toolchain.isDebuggerSupported()) {
            LOG.info("Couldn't find debug-compatible C++ toolchain with name \"Zig\"");
            toolchain = toolchains.getDefaultToolchain();
        }
        if (toolchain == null || !toolchain.isDebuggerSupported()) {
            LOG.info("Couldn't find debug-compatible C++ default toolchain");
            return null;
        }

        CPPToolchains.Toolchain finalToolchain = toolchain;
        return switch (toolchain.getDebuggerKind()) {
            case CUSTOM_GDB, BUNDLED_GDB -> () -> new CLionGDBDriverConfiguration(project, finalToolchain);
            case BUNDLED_LLDB -> () -> new CLionLLDBDriverConfiguration(project, finalToolchain);
        };
    }
}

File: cidr/src/main/kotlin/com/falsepattern/zigbrains/clion/ZigClionDebuggerDriverConfigurationProvider.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2025 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.clion

import com.falsepattern.zigbrains.debugger.ZigDebuggerDriverConfigurationProvider
import com.falsepattern.zigbrains.debugger.settings.ZigDebuggerSettings
import com.intellij.openapi.diagnostic.logger
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.SystemInfo
import com.jetbrains.cidr.cpp.execution.debugger.backend.CLionGDBDriverConfiguration
import com.jetbrains.cidr.cpp.execution.debugger.backend.CLionLLDBDriverConfiguration
import com.jetbrains.cidr.cpp.toolchains.CPPDebugger
import com.jetbrains.cidr.cpp.toolchains.CPPToolchains
import com.jetbrains.cidr.execution.debugger.backend.DebuggerDriverConfiguration

class ZigClionDebuggerDriverConfigurationProvider: ZigDebuggerDriverConfigurationProvider() {
    override suspend fun getDebuggerConfiguration(
        project: Project,
        isElevated: Boolean,
        emulateTerminal: Boolean
    ): DebuggerDriverConfiguration? {
        if (SystemInfo.isWindows)
            return null

        if (!ZigDebuggerSettings.instance.useClion)
            return null

        val toolchains = CPPToolchains.getInstance()
        var toolchain = toolchains.getToolchainByNameOrDefault("Zig")
        if (toolchain == null || !toolchain.isDebuggerSupported) {
            LOG.info("Couldn't find debug-compatible C++ toolchain with name \"Zig\"")
            toolchain = toolchains.defaultToolchain
        }
        if (toolchain == null || !toolchain.isDebuggerSupported) {
            LOG.info("Couldn't find debug-compatible C++ default toolchain")
            return null
        }
        return when(toolchain.debuggerKind) {
            CPPDebugger.Kind.BUNDLED_GDB,
            CPPDebugger.Kind.CUSTOM_GDB -> CLionGDBDriverConfiguration(project, toolchain, isEmulateTerminal = emulateTerminal)
            CPPDebugger.Kind.BUNDLED_LLDB,
            CPPDebugger.Kind.CUSTOM_LLDB -> CLionLLDBDriverConfiguration(project, toolchain, isEmulateTerminal = emulateTerminal)
            CPPDebugger.Kind.CUSTOM_DEBUGGER -> null
        }
    }
}

private val LOG = logger<ZigClionDebuggerDriverConfigurationProvider>()